R/parest.R

```diff
@@ -171,163 +215,244 @@ parestimate.1d.ssa <- function(x, groups, method = c("esprit-ls", "esprit-tls",
+}
+
+# TODO Use a solution of assignment problem
+.simple.assignment <- function(mx) {
+  mx <- as.matrix(mx)
+  stopifnot(ncol(mx) == nrow(mx))
+  d <- nrow(mx)
+  stopifnot(all(mx > -Inf))
+  res <- rep(0, d)
+  for (k in seq_len(d)) {
+    maxij <- which(mx  == max(mx), arr.ind = TRUE)[1, ]
+    res[maxij[1]] <- maxij[2]
+    mx[maxij[1], ] <- -Inf
+    mx[, maxij[2]] <- -Inf
+  }
+
+  res
 }
 
-est_exp_memp_new <- function(Zs, beta = 8) {
-  Z <- (1-beta) * Zs$Zx + beta * Zs$Zy
+.est.exp.memp.new <- function(Zs, beta = 8) {
+  Z <- .matrix.linear.combination(Zs, beta)
   Ze <- eigen(Z)
-  Zxe <- eigen(Zs$Zx)
-  Zye <- eigen(Zs$Zy)
-  Px <- max.col(t(abs(qr.solve(Ze$vectors, Zxe$vectors))))
-  Py <- max.col(t(abs(qr.solve(Ze$vectors, Zxe$vectors))))
+  Zse <- lapply(Zs, eigen, symmetric = FALSE)
+  Ps <- lapply(seq_along(Zs),
+               function(i) .simple.assignment(t(abs(qr.solve(Ze$vectors, Zse[[i]]$vectors)))))
 
-  list(Zxe$values[Px], Zye$values[Py])
+  for (P in Ps) {
+    stopifnot(length(P) == length(unique(P)))
+  }
+
+  lapply(seq_along(Zs),
+         function(i) Zse[[i]]$values[Ps[[i]]])
 }
 
-parestimate.esprit2d <- function(U, L,
-                                 wmask = NULL,
-                                 topology = c(Inf, Inf),
-                                 normalize = c(FALSE, FALSE),
-                                 method = c("esprit-diag-ls", "esprit-diag-tls",
-                                            "esprit-memp-ls", "esprit-memp-tls"),
-                                 beta = 8) {
-  method <- match.arg(method)
-  solve.method <- switch(method,
-                         `esprit-diag-ls`  =, `esprit-memp-ls`  = "ls",
-                         `esprit-diag-tls` =, `esprit-memp-tls` = "tls")
+.parestimate.esprit.nd <- function(U,
+                                   wmask,
+                                   circular,
+                                   normalize,
+                                   dimensions = NULL,
+                                   solve.method = c("ls", "tls"),
+                                   pairing.method = c("diag", "memp"),
+                                   beta = 8) {
+  wmask <- as.array(wmask)
+  d <- dim(wmask)
 
-  Zs <- shift.matrices.2d(U,
-                          L = L,
-                          wmask = wmask,
-                          topology = topology,
-                          solve.method = solve.method)
+  solve.method <- match.arg(solve.method)
+  pairing.method <- match.arg(pairing.method)
 
-  r <- switch(method,
-              `esprit-diag-ls` =, `esprit-diag-tls` = est_exp_2desprit(Zs, beta = beta),
-              `esprit-memp-ls` =, `esprit-memp-tls` = est_exp_memp_new(Zs, beta = beta))
+  if (is.null(dimensions)) {
+    dimensions <- seq_along(d)
+  }
 
-  for (k in 1:2)
+  if (max(dimensions) > length(d)) {
+    stop(sprintf("Some of dimension indices passed exceed the actual number of object dimensions (%d)",
```

Comment:
some of input dimension indices exceed the actual number of object dimensions (%d)

Use stopifnot